data_steward/cdr_cleaner/cleaning_rules/deid/questionnaire_response_id_map.py

```diff
@@ -0,0 +1,153 @@
+"""
+Maps questionnaire_response_ids from the observation table to the research_response_id in the
+_deid_questionnaire_response_map lookup table.
+
+Original Issue: DC-1347
+
+The purpose of this cleaning rule is to create (if it does not already exist) the questionnaire mapping lookup table
+and use that lookup table to remap the questionnaire_response_id in the observation table to the randomly
+generated research_response_id in the _deid_questionnaire_response_map table.
+"""
+
+# Python imports
+import logging
+
+# Project imports
+from utils import pipeline_logging
+from common import OBSERVATION, JINJA_ENV
+from constants.bq_utils import WRITE_TRUNCATE
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+
+LOGGER = logging.getLogger(__name__)
+
+ISSUE_NUMBERS = ['DC1347']
+
+# Creates _deid_questionnaire_response_map lookup table and populates with the questionnaire_response_id's
+# from the observation table as well as randomly generates values for hte research_response_id column
+LOOKUP_TABLE_CREATION_QUERY = JINJA_ENV.from_string("""
+CREATE TABLE IF NOT EXISTS `{{project_id}}.{{shared_sandbox_id}}._deid_questionnaire_response_map` 
+(questionnaire_response_id INT64, research_response_id INT64)
+OPTIONS (description='lookup table for questionnaire response ids') AS
+SELECT DISTINCT questionnaire_response_id AS questionnaire_response_id, ROW_NUMBER() OVER(ORDER BY GENERATE_UUID())
```

Comment:
this is neat/good to know, I had no idea there was this function to use in BQ to create a unique ID!

Reply:
In the original code a lower bound and upper bound for the random number is set, do you know if this needs to be taken into consideration through the query as well? If I remember correctly the `lower_bound` when the pipeline is run was set to 1m.

```# create the deid table.  set upper and lower bounds of the research_id array
    records = table.shape[0]
    upper_bound = lower_bound + (10 * records)
    qr_id_list = table['questionnaire_response_id'].tolist()
    map_table = pd.DataFrame({"questionnaire_response_id": qr_id_list})

    # generate random research_questionnaire_response_ids
    research_id_array = np.random.choice(np.arange(lower_bound, upper_bound),
                                         records,
                                         replace=False)```

Reply:
Yes, Chao helped me create this. `GENERATE_UUID` will create a 16 digit alpha numeric string but combined with the other statements, we got it to create a numeric unique ID 🥳🥳

Reply:
> In the original code a lower bound and upper bound for the random number is set, do you know if this needs to be taken into consideration through the query as well? If I remember correctly the `lower_bound` when the pipeline is run was set to 1m.

I will see if I can modify it so that it'll start on 1m